osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/DefaultSpinnerDisc.cs

```diff
@@ -0,0 +1,170 @@
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            AccentColour = normalColour
+                        },
+                    }
+                },
+                centre = new SpinnerCentreLayer
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                },
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            drawableSpinner.RotationTracker.Complete.BindValueChanged(complete => updateComplete(complete.NewValue, 200));
+            drawableSpinner.State.BindValueChanged(updateStateTransforms, true);
+        }
+
+        private void updateStateTransforms(ValueChangedEvent<ArmedState> state)
+        {
+            centre.ScaleTo(0);
+            mainContainer.ScaleTo(0);
+
+            using (BeginAbsoluteSequence(spinner.StartTime - spinner.TimePreempt / 2, true))
+            {
+                // constant ambient rotation to give the spinner "spinning" character.
+                this.RotateTo((float)(25 * spinner.Duration / 2000), spinner.TimePreempt + spinner.Duration);
+
+                centre.ScaleTo(0.3f, spinner.TimePreempt / 4, Easing.OutQuint);
+                mainContainer.ScaleTo(0.2f, spinner.TimePreempt / 4, Easing.OutQuint);
+
+                using (BeginDelayedSequence(spinner.TimePreempt / 2, true))
+                {
+                    centre.ScaleTo(0.5f, spinner.TimePreempt / 2, Easing.OutQuint);
+                    mainContainer.ScaleTo(1, spinner.TimePreempt / 2, Easing.OutQuint);
+                }
+            }
+
+            // transforms we have from completing the spinner will be rolled back, so reapply immediately.
+            updateComplete(state.NewValue == ArmedState.Hit, 0);
+
+            using (BeginDelayedSequence(spinner.Duration, true))
+            {
+                switch (state.NewValue)
+                {
+                    case ArmedState.Hit:
+                        this.ScaleTo(Scale * 1.2f, 320, Easing.Out);
+                        this.RotateTo(mainContainer.Rotation + 180, 320);
+                        break;
+
+                    case ArmedState.Miss:
+                        this.ScaleTo(Scale * 0.8f, 320, Easing.In);
+                        break;
+                }
+            }
+        }
+
+        private void updateComplete(bool complete, double duration)
+        {
+            var colour = complete ? completeColour : normalColour;
+
+            ticks.FadeAccent(colour.Darken(1), duration);
+            fill.FadeAccent(colour.Darken(1), duration);
+
+            background.FadeAccent(colour, duration);
+            centre.FadeAccent(colour, duration);
+        }
+
+        private bool updateCompleteTick() => completeTick != (completeTick = (int)(drawableSpinner.RotationTracker.CumulativeRotation / 360));
```

Comment:
wow, this, uh... I thought I had brain damage for a little bit when reading this as to how this can ever be true (but it can!).

[analogous, simpler example for the curious](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2CAbJmAHbCYAOA3MaSyTBXgBQCUhbpEvUwBeTOmYpBAXwECMZSgCMA9ioiYAsr35TBpGAHYGmAIRiuwsQFYekmXL3s4cXfoMIAnF2087AwQ5vX38nfSCfXlD3di9Iv3sYiJDE92So1MxZFGkgA===). CLR implicitly stores the old value of the variable onto the bottom of the stack before proceeding with execution (below is the explanation for the first run that returns `true`; every other run after this will return `false`):

```
                                        p = 3   stack: (empty)
        IL_0000: ldsfld int32 C::p      p = 3   stack: 3
        IL_0005: ldc.i4.5               p = 3   stack: 3 5
        IL_0006: dup                    p = 3   stack: 3 5 5
        IL_0007: stsfld int32 C::p      p = 5   stack: 3 5
        IL_000c: ceq                    p = 5   stack: 0
        IL_000e: ldc.i4.0               p = 5   stack: 0 0
        IL_000f: ceq                    p = 5   stack: 1
        IL_0011: ret                    p = 5   function returns 1
```

so this effectively translates to, "store the value on the right side of the assignment to `completeTick`, then compare the value assigned to the old value and return the result of that comparison".

I don't know about you but I've literally never seen this pattern before. might be worth an inline comment or something? it read to me like a bug at first.

Reply:
i'll split it out to read better